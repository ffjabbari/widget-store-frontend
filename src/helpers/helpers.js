import { Redirect } from 'react-router-dom';
import { motion } from 'framer-motion';

/**
 * Given a path return the "root" argument.
 * For example the "root" of /shop/product/206 would be "/shop".
 *
 * @param {string} path
 */

export const getPathRoot = path => {
  return path.match(/^(\/\w*)\/?.*$/)[1];
};

export const FramerRedirect = props => (
  <motion.div exit="undefined">
    <Redirect {...props} />
  </motion.div>
);

export const animateVariant = (setterFunct, delay, callback, args = []) => {
  // if animation state is false set to 'default' else 'active'
  setterFunct(state => (state === false ? 'default' : 'active'));
  setTimeout(() => {
    setterFunct('default');
    if (callback) callback(...args);
  }, delay);
};
