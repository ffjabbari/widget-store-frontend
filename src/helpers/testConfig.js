/** Test data setup/teardown helper hooks */

export const TEST_DATA = {};

export const populateTestDataHook = testData => {
  testData.product = {
    byline: 'Latest generation smart televison 2021',
    date_added: '2020-12-29T08:00:00.000Z',
    description:
      'Lorem ipsum dolor sit, amet consectetur adipisicing elit. Nihil, voluptates enim nam voluptatum, quia doloribus quis sequi, porro consequatur excepturi eos error voluptas nobis modi atque magni commodi amet placeat.',
    discount: '0.00',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Samsung XG-900 55 inch LCD TV',
    price: '400.40',
    product_id: 1,
    quantity: 100,
    rating: null,
  };

  testData.product2 = {
    byline: 'A good remote.',
    discount: '0.15',
    image_url:
      'https://www.lg.com/sg/images/tvs/md05271244/gallery/LG-43LH600T-L.jpg',
    name: 'Logitech Singularity Universal Remote',
    price: '40.20',
    product_id: 2,
    rating: null,
    quantity: 1,
  };

  testData.userProfileData = {
    data: {
      user: {
        user_id: 1,
        username: 'widgetLord',
        email: 'foo@gmail.com',
        first_name: 'bob',
        last_name: 'bobert',
        address: '123 Hill Ave.',
        address_line2: null,
        city: 'Big City',
        state: 'Texas',
        postal_code: '33333',
        phone_number: null,
        avatar_url: null,
        orders: [
          {
            order_id: 1,
            order_date: '2021-01-24T08:00:00.000Z',
            status: 'Pending',
          },
        ],
      },
    },
  };

  testData.CustomerInfoFormsValues = {
    first_name: 'Test',
    last_name: 'User',
    email: 'foo@gmail.com',
    address: '123 Main St',
    address_line2: '',
    city: 'Big City',
    state: 'TX',
    postal_code: '39488',
    phone_number: '(555) 555-5555',
    set_default: false,
  };

  testData.cart = {
    items: {
      2: {
        byline: 'The only remote you will need',
        date_added: '2021-01-24T08:00:00.000Z',
        description: 'Lorem ipsum dolor sit, placeat.',
        discount: '0.15',
        image_url:
          'https://cnet3.cbsistatic.com/img/SB-Lc9lni02k-GQeFZlVGII6cpg=/1070x602/2010/08/03/57369690-bb77-11e2-8a8e-0291187978f3/33981226-2-1333-OVR-1.jpg',
        name: 'Logitech Singularity Universal Remote',
        price: '40.20',
        product_id: 2,
        quantity: 2,
        rating: null,
      },
    },
    numCartItems: 1,
    subtotal: '34.17',
  };
};
