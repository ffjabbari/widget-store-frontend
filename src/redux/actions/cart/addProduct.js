/** Add_Product action creator. */

import { ADD_PRODUCT } from '../types';

function addProduct(payload) {
  return { type: ADD_PRODUCT, payload };
}
export default addProduct;
