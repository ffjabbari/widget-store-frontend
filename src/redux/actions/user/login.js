/** Log user in action creator. */

import { LOG_IN } from '../types';

function login(user, token) {
  return { type: LOG_IN, payload: { user, token } };
}

export default login;
