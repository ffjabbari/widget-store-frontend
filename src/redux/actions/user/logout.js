/** Log user out action creator. */

import { LOG_OUT } from '../types';

function logout() {
  return { type: LOG_OUT };
}

export default logout;
