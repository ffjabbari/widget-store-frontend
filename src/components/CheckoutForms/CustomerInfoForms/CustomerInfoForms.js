import { useState, useEffect, useRef } from 'react';
import { shallowEqual, useSelector } from 'react-redux';
import { NavLink } from 'react-router-dom';
import * as Yup from 'yup';
import { Button } from 'reactstrap';

// local imports
import { asyncAPIRequest } from '../../../hooks/apiHook';
import ShippingAddressForm from './ShippingAddressForm';
import CustomerAddressForm from './CustomerAddressForm';
import AddressBox from '../AddressBox/AddressBox';
import PaymentOptions from '../../common/PaymentOptions/PaymentOptions';

/**
 * Customer info forms for checkout stage 1.
 *
 */
const CustomerInfoForms = ({
  orderData,
  setOrderData,
  goTo1,
  goTo2,
  getShippingCosts,
  subtotal,
  setCustomerCheckmark,
}) => {
  const { user, token } = useSelector(
    state => ({
      user: state.user.user,
      token: state.user.token,
    }),
    shallowEqual
  );
  // vars for API request to user profile
  const [responseError, setResponseError] = useState(null);
  const [loadingUser, setLoadingUser] = useState(true);
  // flag vars for customer and shipping forms display control, forms reset
  const [formDisabled, setFormDisabled] = useState(false);
  const [shippingFormDisabled, setShippingFormDisabled] = useState(true);
  const [formReset, setFormReset] = useState(false);

  // Note orginal user when page loads
  const userOriginal = useRef(token);
  const defaultAddress = useRef(false);
  const FORM_DATA = useRef(DEFAULT_CUST_FORM_DATA);

  // get user profile data if user is logged in
  useEffect(() => {
    async function getUser() {
      try {
        const resp = await asyncAPIRequest(
          `/users/${user.username}`,
          'get',
          null,
          {
            _token: token,
          }
        );
        if (resp.error) setResponseError(resp.error.response.data.message);
        else {
          // parse relevant data
          const userData = getCustomerDataFromResponse(resp);

          // update default form state to show user info on form
          FORM_DATA.current = {
            ...FORM_DATA.current,
            ...userData,
          };
          // determine if enough info to have a default shipping address
          defaultAddress.current = enoughInfoForDefaultAddress(userData);

          // disable form is sufficient shipping address data is available
          if (defaultAddress.current) setFormDisabled(true);

          setLoadingUser(false);
        }
      } catch (error) {
        console.log(error);
        setResponseError(error.message);
      }
    }
    // only attempt getting user info if user is logged in
    if (token) getUser();
  }, [user, token]);

  // if user logs in or out reset forms and data
  useEffect(() => {
    let reset;
    async function userChangeCheck() {
      if (token !== userOriginal.current) {
        // remove customer data and shipping data.
        setOrderData(orderData => ({
          ...orderData,
          customer: null,
          shipping: false,
          shippingAddress: null,
          tax: null,
        }));
        // set new user token reference
        userOriginal.current = token;
        // reset variables
        defaultAddress.current = false;
        setFormDisabled(false);
        setShippingFormDisabled(true);

        // Forms Reset-
        // Set form data to default if user is not signed in.
        if (token === null) FORM_DATA.current = DEFAULT_CUST_FORM_DATA;
        // reset forms by remounting forms.
        setFormReset(true);
        // reset = await new Promise(res => setTimeout(res, 100));
        // setFormReset(false);
        reset = setTimeout(() => {
          setFormReset(false);
        }, 100);

        goTo1();
      }
    }
    userChangeCheck();

    return () => {
      clearTimeout(reset);
    };
  }, [token, setOrderData, goTo1]);

  // if user is logged in but user profile data has not loaded yet return loading...
  if (token && loadingUser) return <h1>Loading...</h1>;

  const editCustomerInfoButton = (
    <Button
      onClick={() => {
        setShippingFormDisabled(true);
        setFormDisabled(formDisabled => !formDisabled);
      }}
      color="primary"
      className="mb-4"
      size="sm"
      outline
    >
      Edit Customer Info
    </Button>
  );

  const editShipAddressButton = (
    <Button
      onClick={() => {
        setFormDisabled(true);
        setShippingFormDisabled(shippigFormDisabled => !shippigFormDisabled);
      }}
      color="primary"
      className="mb-4"
      size="sm"
      outline
    >
      Edit Shipping Info
    </Button>
  );

  const customerAddressFormProps = {
    orderData,
    setOrderData,
    FORM_DATA,
    shippingFormDisabled,
    user,
    token,
    formDisabled,
    responseError,
    setResponseError,
    defaultAddress,
    setFormDisabled,
    goTo2,
    getShippingCosts,
    subtotal,
    setCustomerCheckmark,
  };

  return (
    <div className="CustomerInfoForms px-1 px-md-4 pt-2 pb-3 bg-light text-dark rounded border">
      {/* If user not logged in show login button */}
      {!user && (
        <div>
          <NavLink
            to="/checkout/login"
            className="btn btn-info btn-sm mb-3 mt-1"
          >
            Login / Create Account
          </NavLink>
          <span className="pl-3">
            <i>- or checkout as guest below</i>
          </span>
        </div>
      )}
      {/* Show address boxes if there was enough info for a default address*/}
      {defaultAddress.current && (
        <div className="d-flex justify-content-around">
          <AddressBox
            title={'Customer Info'}
            customer={FORM_DATA.current}
            button={editCustomerInfoButton}
            className="mx-0"
          />
          <AddressBox
            title={'Shipping Address'}
            customer={orderData.shippingAddress || FORM_DATA.current}
            isShippingAddress="true"
            button={editShipAddressButton}
            className="mx-0"
          />
        </div>
      )}
      {!formReset && (
        <>
          <CustomerAddressForm {...customerAddressFormProps} />
          <ShippingAddressForm
            setOrderData={setOrderData}
            FORM_DATA={FORM_DATA}
            setShippingFormDisabled={setShippingFormDisabled}
            shippingFormDisabled={shippingFormDisabled}
          />
        </>
      )}
      <PaymentOptions />
    </div>
  );
};

const DEFAULT_CUST_FORM_DATA = {
  first_name: '',
  last_name: '',
  email: '',
  address: '',
  address_line2: '',
  city: '',
  state: '',
  postal_code: '',
  phone_number: '',
};

/** Form validation logic */
export const customerInfoSchema = Yup.object({
  first_name: Yup.string()
    .max(55, 'Must be 55 characters or less')
    .required('Required'),
  last_name: Yup.string()
    .max(55, 'Must be 55 characters or less')
    .required('Required'),
  email: Yup.string().email('Invalid email address').required('Required'),
  address: Yup.string()
    .max(255, 'Must be 255 characters or less')
    .min(8, 'Must be 8 charaters or more')
    .required('Required'),
  address_line2: Yup.string().max(255, 'Must be 255 characters or less'),
  city: Yup.string()
    .max(55, 'Must be 255 characters or less')
    .min(2, 'Must be 2 charaters or more')
    .required('Required'),
  state: Yup.string()
    .max(55, 'Must be 255 characters or less')
    .min(2, 'Must be 2 charaters or more')
    .required('Required'),
  postal_code: Yup.string()
    .matches(/^\d{5}(-?\d{4})?$/, 'Must be 5 or 9 digit postal code')
    .required('Required'),
  phone_number: Yup.string().matches(
    /^(?:\+?(\d{1,3})[- .]?)?[(]?(\d{3})(?:(?:\) )?|[-. )]?)(\d{3})[-. ]?(\d{4})$/,
    'Must be valid phone number'
  ),
});

/**
 * Create a userData object with only the relevant properties to the
 * customer info form. Created from the user profile API response object.
 *
 * @param {Object} resp
 *
 * returns: Object
 */

const getCustomerDataFromResponse = resp => {
  const userData = { ...resp.data.user };
  // delete unneccasary properties for customer info form

  delete userData['orders'];
  delete userData['avatar_url'];
  delete userData['username'];
  // change null values to empty string '' value
  Object.entries(userData).forEach(([key, value]) => {
    if (value === null) userData[key] = '';
  });
  return userData;
};

/**
 * @param {Object} userData
 *
 * Determine if there is enough user data on file to have
 * a default shipping address to use.
 */

const enoughInfoForDefaultAddress = userData => {
  return Object.entries(userData).every(([key, value]) => {
    // ignore the keys optional to having a complete shipping address
    if (['address_line2', 'phone_number'].includes(key)) return true;
    // address data is required
    else return !!value;
  });
};

export default CustomerInfoForms;
