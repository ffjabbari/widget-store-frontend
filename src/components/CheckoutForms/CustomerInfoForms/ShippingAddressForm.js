import { Formik, Form, Field, ErrorMessage } from 'formik';
import { Button, FormGroup, Label, Col, Row, Collapse } from 'reactstrap';

// local imports
import { customerInfoSchema } from './CustomerInfoForms';

const ShippingForm = ({
  setOrderData,
  FORM_DATA,
  setShippingFormDisabled,
  shippingFormDisabled,
}) => {
  // handle update shipping address
  const handleSubmitShipping = values => {
    setOrderData(orderData => ({
      ...orderData,
      shippingAddress: {
        ...values,
        phone_number:
          orderData.customer?.phone_number || FORM_DATA.current.phone_number,
      },
      shipping: false,
    }));
    setShippingFormDisabled(true);
  };

  return (
    <Formik
      initialValues={FORM_DATA.current}
      validationSchema={customerInfoSchema}
      onSubmit={(values, { setSubmitting }) => {
        handleSubmitShipping(values);
        setSubmitting(false);
      }}
    >
      {formik => (
        <Form className="text-left ShippingAddressForm">
          <Collapse isOpen={!shippingFormDisabled}>
            <h6 className="font-weight-bold">Shipping Address</h6>
            <Row form>
              <Col sm={6}>
                <FormGroup>
                  <Label htmlFor="first_name_shipping">First Name</Label>
                  <Field
                    type="text"
                    name="first_name"
                    id="first_name_shipping"
                    className={`form-control form-control-sm ${
                      formik.touched.first_name_shipping &&
                      formik.errors.first_name_shipping &&
                      'is-invalid'
                    }`}
                    disabled={shippingFormDisabled}
                  />
                  <ErrorMessage
                    name="first_name"
                    component="div"
                    className="text-danger mt-1 text-sm"
                  />
                </FormGroup>
              </Col>
              <Col sm={6}>
                <FormGroup>
                  <Label htmlFor="last_name_shipping">Last Name</Label>
                  <Field
                    type="text"
                    name="last_name"
                    id="last_name_shipping"
                    className={`form-control form-control-sm ${
                      formik.touched.last_name &&
                      formik.errors.last_name &&
                      'is-invalid'
                    }`}
                    disabled={shippingFormDisabled}
                  />
                  <ErrorMessage
                    name="last_name_shipping"
                    component="div"
                    className="text-danger mt-1 text-sm"
                  />
                </FormGroup>
              </Col>
            </Row>

            <FormGroup>
              <Label htmlFor="address_shipping">Street Address</Label>
              <Field
                type="text"
                name="address"
                id="address_shipping"
                className={`form-control form-control-sm ${
                  formik.touched.address &&
                  formik.errors.address &&
                  'is-invalid'
                }`}
                disabled={shippingFormDisabled}
              />
              <ErrorMessage
                name="address"
                component="div"
                className="text-danger mt-1 text-sm"
              />
            </FormGroup>
            <Row form>
              <Col sm={6}>
                <FormGroup>
                  <Label htmlFor="address_line2_shipping">
                    Address Line Two
                  </Label>
                  <Field
                    type="text"
                    name="address_line2"
                    id="address_line2_shipping"
                    className={`form-control form-control-sm ${
                      formik.touched.address_line2 &&
                      formik.errors.address_line2 &&
                      'is-invalid'
                    }`}
                    disabled={shippingFormDisabled}
                  />
                  <ErrorMessage
                    name="address_line2"
                    component="div"
                    className="text-danger mt-1 text-sm"
                  />
                </FormGroup>
              </Col>
              <Col sm={6}>
                <FormGroup>
                  <Label htmlFor="city_shipping">City</Label>
                  <Field
                    type="text"
                    name="city"
                    id="city_shipping"
                    className={`form-control form-control-sm ${
                      formik.touched.city && formik.errors.city && 'is-invalid'
                    }`}
                    disabled={shippingFormDisabled}
                  />
                  <ErrorMessage
                    name="city"
                    component="div"
                    className="text-danger mt-1 text-sm"
                  />
                </FormGroup>
              </Col>
            </Row>
            <Row form>
              <Col sm={6}>
                <FormGroup>
                  <Label htmlFor="state_shipping">State</Label>
                  <Field
                    type="text"
                    name="state"
                    id="state_shipping"
                    className={`form-control form-control-sm ${
                      formik.touched.state &&
                      formik.errors.state &&
                      'is-invalid'
                    }`}
                    disabled={shippingFormDisabled}
                  />
                  <ErrorMessage
                    name="state"
                    component="div"
                    className="text-danger mt-1 text-sm"
                  />
                </FormGroup>
              </Col>
              <Col sm={6}>
                <FormGroup>
                  <Label htmlFor="postal_code_shipping">Postal Code</Label>
                  <Field
                    type="text"
                    name="postal_code"
                    id="postal_code_shipping"
                    className={`form-control form-control-sm ${
                      formik.touched.postal_code &&
                      formik.errors.postal_code &&
                      'is-invalid'
                    }`}
                    disabled={shippingFormDisabled}
                  />
                  <ErrorMessage
                    name="postal_code"
                    component="div"
                    className="text-danger mt-1 text-sm"
                  />
                </FormGroup>
              </Col>
            </Row>

            <div className="text-right">
              <Button
                type="submit"
                color="primary"
                disabled={formik.isSubmitting}
                className="rounded-pill"
              >
                Update Shipping Address
              </Button>
              <Button
                color="secondary"
                className="ml-2"
                onClick={() =>
                  setShippingFormDisabled(
                    shippigFormDisabled => !shippigFormDisabled
                  )
                }
              >
                Cancel
              </Button>
            </div>
          </Collapse>
        </Form>
      )}
    </Formik>
  );
};

export default ShippingForm;
