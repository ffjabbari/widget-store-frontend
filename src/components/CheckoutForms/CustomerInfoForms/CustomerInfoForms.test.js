import { act, fireEvent } from '@testing-library/react';
import axios from 'axios';

// local imports
import { renderWithStore } from '../../../helpers/testHelpers';
import { TEST_DATA, populateTestDataHook } from '../../../helpers/testConfig';
import { testStore } from '../../../redux/store/reduxStore';
import login from '../../../redux/actions/user/login';
import logout from '../../../redux/actions/user/logout';
import addProduct from '../../../redux/actions/cart/addProduct';
import CustomerInfoForms from './CustomerInfoForms';

// mock axios, setOrderData, and goTo2
jest.mock('axios');
const setOrderData = jest.fn();
const goTo1 = jest.fn();
const goTo2 = jest.fn();
const getShippingCosts = jest.fn();
const setCustomerCheckmark = jest.fn();

// label regexs and test values for form inputs testing
const inputs = [
  [/first name/i, 'bob'],
  [/last name/i, 'bobert'],
  [/email/i, 'abc@def.gh'],
  [/street address/i, '123 Hill Ave'],
  [/city/i, 'Big City'],
  [/state/i, 'Texas'],
  [/postal code/i, '55555'],
];

beforeAll(() => {
  populateTestDataHook(TEST_DATA);
  testStore.dispatch(addProduct(TEST_DATA.product));
});

afterEach(() => {
  jest.resetAllMocks();
});

test('CustomerInfoForms should render', () => {
  renderWithStore(
    <CustomerInfoForms
      orderData={{ shipping: false }}
      setOrderData={setOrderData}
      goTo1={goTo1}
      goTo2={goTo2}
      getShippingCosts={getShippingCosts}
      subtotal={TEST_DATA.cart.subtotal}
      setCustomerCheckmark={setCustomerCheckmark}
    />
  );
});

test('CustomerInfoForms snapshot', async () => {
  const { asFragment } = renderWithStore(
    <CustomerInfoForms
      orderData={{ shipping: false }}
      setOrderData={setOrderData}
      goTo1={goTo1}
      goTo2={goTo2}
      getShippingCosts={getShippingCosts}
      subtotal={TEST_DATA.cart.subtotal}
      setCustomerCheckmark={setCustomerCheckmark}
    />
  );
  expect(asFragment()).toMatchSnapshot();
});

test('shoud add customer data to orderData with submit success', async () => {
  const { getAllByLabelText, getByRole } = renderWithStore(
    <CustomerInfoForms
      orderData={{ shipping: false }}
      setOrderData={setOrderData}
      goTo1={goTo1}
      goTo2={goTo2}
      getShippingCosts={getShippingCosts}
      subtotal={TEST_DATA.cart.subtotal}
      setCustomerCheckmark={setCustomerCheckmark}
    />
  );

  // add values to all inputs
  await act(async () => {
    inputs.forEach(async ([re, text]) => {
      const input = getAllByLabelText(re)[0];
      fireEvent.change(input, { target: { value: text } });
    });
  });

  const nextButton = getByRole('button', {
    name: /next/i,
  });
  await act(async () => {
    fireEvent.click(nextButton);
  });
  expect(setOrderData).toHaveBeenCalled();
  expect(getShippingCosts).toHaveBeenCalled();
  expect(getShippingCosts.mock.calls.length).toBe(1);
  expect(getShippingCosts.mock.calls[0][0].shippingAddress.address).toBe(
    '123 Hill Ave'
  );
  expect(getShippingCosts.mock.calls[0][0].shippingAddress.city).toBe(
    'Big City'
  );
  expect(getShippingCosts.mock.calls[0][0].shippingAddress.state).toBe('Texas');
  expect(getShippingCosts.mock.calls[0][0].shippingAddress.postal_code).toBe(
    '55555'
  );

  // call funct that was argument to setOrderData with blank "orderData" object
  const orderData = setOrderData.mock.calls[0][0]({});
  expect(orderData).toHaveProperty('customer');
  expect(orderData.customer.first_name).toBe('bob');
  expect(Object.keys(orderData.customer)).toHaveLength(9);
});

test('shoud auto add customer data if user is signed in', async () => {
  testStore.dispatch(login({ username: 'user' }, 'token'));
  axios.mockResolvedValue(TEST_DATA.userProfileData);

  let getByRole;
  await act(async () => {
    const resp = renderWithStore(
      <CustomerInfoForms
        orderData={{ shipping: false }}
        setOrderData={setOrderData}
        goTo1={goTo1}
        goTo2={goTo2}
        getShippingCosts={getShippingCosts}
        subtotal={TEST_DATA.cart.subtotal}
        setCustomerCheckmark={setCustomerCheckmark}
      />
    );
    getByRole = resp.getByRole;
  });

  const nextButton = getByRole('button', {
    name: /next/i,
  });
  await act(async () => {
    fireEvent.click(nextButton);
  });

  expect(axios).toHaveBeenCalled();
  expect(axios.mock.calls.length).toBe(1);
  expect(axios.mock.calls[0][0].url).toBe('/users/user');
  expect(axios.mock.calls[0][0].method).toBe('get');
  expect(axios.mock.calls[0][0].params._token).toBe('token');

  expect(setOrderData).toHaveBeenCalled();
  expect(setOrderData.mock.calls.length).toBe(1);
  expect(getShippingCosts).toHaveBeenCalled();
  expect(getShippingCosts.mock.calls.length).toBe(1);

  // expect orderData to have been populated
  const orderData = setOrderData.mock.calls[0][0]({});
  expect(orderData).toHaveProperty('customer');
  expect(orderData.customer.first_name).toBe('bob');
  expect(Object.keys(orderData.customer)).toHaveLength(10);

  testStore.dispatch(logout());
});

test('shoud not update orderData with incomplete form', async () => {
  const { getAllByLabelText, getByRole, getByText } = renderWithStore(
    <CustomerInfoForms
      orderData={{ shipping: false }}
      setOrderData={setOrderData}
      goTo1={goTo1}
      goTo2={goTo2}
      getShippingCosts={getShippingCosts}
      subtotal={TEST_DATA.cart.subtotal}
      setCustomerCheckmark={setCustomerCheckmark}
    />
  );

  // choose random index for a required input to skip
  const skip = Math.floor(Math.random() * inputs.length);

  // add values to inputs for all but one random input
  await act(async () => {
    inputs.forEach(async ([re, text], idx) => {
      if (idx === skip) return;
      const input = getAllByLabelText(re)[0];
      fireEvent.change(input, { target: { value: text } });
    });
  });

  const nextButton = getByRole('button', {
    name: /next/i,
  });
  await act(async () => {
    fireEvent.click(nextButton);
  });

  // expect one input on form to show "required"
  const invalidInput = getByText('Required');
  expect(invalidInput).toBeInTheDocument();
});

test('shoud call update profile when user signed in and update checked', async () => {
  testStore.dispatch(login({ username: 'user' }, 'token'));
  axios.mockResolvedValue(TEST_DATA.userProfileData);

  let getByRole;
  await act(async () => {
    const resp = renderWithStore(
      <CustomerInfoForms
        orderData={{ shipping: false }}
        setOrderData={setOrderData}
        goTo1={goTo1}
        goTo2={goTo2}
        getShippingCosts={getShippingCosts}
        subtotal={TEST_DATA.cart.subtotal}
        setCustomerCheckmark={setCustomerCheckmark}
      />
    );
    getByRole = resp.getByRole;
  });

  // click edit info
  await act(async () => {
    const editButton = getByRole('button', { name: 'Edit Customer Info' });
    fireEvent.click(editButton);
  });

  axios.mockResolvedValue({
    data: { ok: 'ok' },
  });

  // update user data
  const updateButton = getByRole('button', {
    name: /Update Customer Info/i,
  });
  await act(async () => {
    fireEvent.click(updateButton);
  });

  expect(axios).toHaveBeenCalled();
  expect(axios.mock.calls.length).toBe(2);
  expect(setOrderData).toHaveBeenCalled();
  // expect(goTo2).not.toHaveBeenCalled();

  expect(axios.mock.calls[1][0].url).toBe('/users/user');
  expect(axios.mock.calls[1][0].method).toBe('patch');
  expect(axios.mock.calls[1][0].params._token).toBe('token');
  expect(axios.mock.calls[1][0].data.first_name).toBe('bob');

  testStore.dispatch(logout());
});

test('shoud show error message with profile update failure', async () => {
  testStore.dispatch(login({ username: 'user' }, 'token'));
  axios.mockResolvedValue(TEST_DATA.userProfileData);

  let getByRole;
  await act(async () => {
    const resp = renderWithStore(
      <CustomerInfoForms
        orderData={{ shipping: false }}
        setOrderData={setOrderData}
        goTo1={goTo1}
        goTo2={goTo2}
        getShippingCosts={getShippingCosts}
        subtotal={TEST_DATA.cart.subtotal}
        setCustomerCheckmark={setCustomerCheckmark}
      />
    );
    getByRole = resp.getByRole;
  });

  // click edit info
  await act(async () => {
    const editButton = getByRole('button', { name: 'Edit Customer Info' });
    fireEvent.click(editButton);
  });

  axios.mockResolvedValue({
    error: { response: { data: { message: 'error message' } } },
  });

  const updateButton = getByRole('button', {
    name: /update customer info/i,
  });
  await act(async () => {
    fireEvent.click(updateButton);
  });

  expect(axios).toHaveBeenCalled();
  expect(axios.mock.calls.length).toBe(2);

  const errorMessage = getByRole('alert');
  expect(errorMessage).toBeInTheDocument();
  expect(errorMessage.textContent).toBe('error message');

  testStore.dispatch(logout());
});
