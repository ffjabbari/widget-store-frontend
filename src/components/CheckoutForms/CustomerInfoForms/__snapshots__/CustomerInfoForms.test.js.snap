// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomerInfoForms snapshot 1`] = `
<DocumentFragment>
  <div
    class="CustomerInfoForms px-1 px-md-4 pt-2 pb-3 bg-light text-dark rounded"
  >
    <div>
      <a
        class="btn btn-info btn-sm mb-3 mt-1"
        href="/checkout/login"
      >
        Login / Create Account
      </a>
      <span
        class="pl-3"
      >
        <i>
          - or checkout as guest below
        </i>
      </span>
    </div>
    <form
      action="#"
      class="text-left CustomerInfoForm"
    >
      <div
        class="collapse show"
      >
        <h6
          class="text-info"
        >
          Customer Info
        </h6>
        <div
          class="form-row"
        >
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="first_name"
              >
                First Name
              </label>
              <input
                autocomplete="given-name"
                class="form-control form-control-sm undefined"
                id="first_name"
                name="first_name"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="last_name"
              >
                Last Name
              </label>
              <input
                autocomplete="family-name"
                class="form-control form-control-sm undefined"
                id="last_name"
                name="last_name"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="form-row"
        >
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="email"
              >
                Email
              </label>
              <input
                autocomplete="email"
                class="form-control form-control-sm undefined"
                id="email"
                name="email"
                type="email"
                value=""
              />
            </div>
          </div>
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="phone_number"
              >
                Phone Number
              </label>
              <input
                autocomplete="tel-national"
                class="form-control form-control-sm undefined"
                id="phone_number"
                name="phone_number"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            class=""
            for="address"
          >
            Street Address
          </label>
          <input
            autocomplete="address-line1"
            class="form-control form-control-sm undefined"
            id="address"
            name="address"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-row"
        >
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="address_line2"
              >
                Address Line Two
              </label>
              <input
                autocomplete="address-line2"
                class="form-control form-control-sm undefined"
                id="address_line2"
                name="address_line2"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="city"
              >
                City
              </label>
              <input
                autocomplete="address-level2"
                class="form-control form-control-sm undefined"
                id="city"
                name="city"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="form-row"
        >
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="state"
              >
                State
              </label>
              <input
                autocomplete="address-level1"
                class="form-control form-control-sm undefined"
                id="state"
                name="state"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="postal_code"
              >
                Postal Code
              </label>
              <input
                autocomplete="postal-code"
                class="form-control form-control-sm undefined"
                id="postal_code"
                name="postal_code"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <button
        class="form-control rounded-pill mt-0 mt-lg-4 btn btn-primary"
        type="submit"
      >
        Next
      </button>
    </form>
    <form
      action="#"
      class="text-left ShippingAddressForm"
    >
      <div
        class="collapse"
      >
        <h6
          class="text-info"
        >
          Shipping Address
        </h6>
        <div
          class="form-row"
        >
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="first_name_shipping"
              >
                First Name
              </label>
              <input
                class="form-control form-control-sm undefined"
                disabled=""
                id="first_name_shipping"
                name="first_name"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="last_name_shipping"
              >
                Last Name
              </label>
              <input
                class="form-control form-control-sm undefined"
                disabled=""
                id="last_name_shipping"
                name="last_name"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            class=""
            for="address_shipping"
          >
            Street Address
          </label>
          <input
            class="form-control form-control-sm undefined"
            disabled=""
            id="address_shipping"
            name="address"
            type="text"
            value=""
          />
        </div>
        <div
          class="form-row"
        >
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="address_line2_shipping"
              >
                Address Line Two
              </label>
              <input
                class="form-control form-control-sm undefined"
                disabled=""
                id="address_line2_shipping"
                name="address_line2"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="city_shipping"
              >
                City
              </label>
              <input
                class="form-control form-control-sm undefined"
                disabled=""
                id="city_shipping"
                name="city"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="form-row"
        >
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="state_shipping"
              >
                State
              </label>
              <input
                class="form-control form-control-sm undefined"
                disabled=""
                id="state_shipping"
                name="state"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="col-sm-6"
          >
            <div
              class="form-group"
            >
              <label
                class=""
                for="postal_code_shipping"
              >
                Postal Code
              </label>
              <input
                class="form-control form-control-sm undefined"
                disabled=""
                id="postal_code_shipping"
                name="postal_code"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="text-right"
        >
          <button
            class="rounded-pill btn btn-primary"
            type="submit"
          >
            Update Shipping Address
          </button>
          <button
            class="ml-2 btn btn-secondary"
            type="button"
          >
            Cancel
          </button>
        </div>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
